src/test/java/HttpServletResponseFake.java

```diff
@@ -0,0 +1,224 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Collection;
+import java.util.Locale;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletResponse;
+
+/** Fake version of HttpServletResponse used for testing */
+public class HttpServletResponseFake implements HttpServletResponse {
```

Comment:
You could avoid all the unsupported method implementations, by using Mockito to delegate calls to your object: https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#27

- This object doesn't implement `HttpServletResponse`, and only declares the methods you want to support.
- Mockito is used to create a mock and forward all calls to your object.
  - The mock will implement `HttpServeletResponse` and all its methods.
- Methods that you have not implemented will produce some sort of error.

Reply:
This is a really cool tool, thanks for pointing this out! 

However, I don't know if this will work for this implementation, mainly due to `stringWriter`. If I use `response = Mockito.mock(HttpServletResponse.class, AdditionalAnswers.delgateTo(HttpServletResponseFake.class));` all methods called on that mock have to be defined in the fake. However, `HttpServletResponse` out of the box does not contain a `getStringWriter()` method, which means if I used this mock I have no way of getting the printed response. This setting also cannot be overriden (saying `Mockito.when(response.getWriter()).thenReturn(stringWriter)` does not work). 

I could just use a regular mock and do `Mockito.when(methodOne()).thenReturn(HttpServletResponseFake.methodOne())` or something along those lines, but that is going to get really messy, especially when the mocked methods take in parameters. Overall, while verbose, it might be cleaner to just create the fake.